Make c10d_functional.wait_tensor effectful ORDERED op

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132755

Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @desertfire @chauhang @aakhundov @XilunWu @ColinPeppler

IvanKobzarev committed:
Make c10d_functional.wait_tensor effectful ORDERED op

[ghstack-poisoned]

fegin commented on torch/_higher_order_ops/effects.py:
```diff
@@ -26,6 +26,7 @@ class _EffectType(Enum):
 SIDE_EFFECTS: Dict[OpType, _EffectType] = {
     torch.ops.aten._print.default: _EffectType.ORDERED,
     call_torchbind: _EffectType.ORDERED,
+    torch.ops._c10d_functional.wait_tensor.default: _EffectType.ORDERED,
```
We should also delete the two lines here, https://github.com/pytorch/pytorch/blob/main/torch/distributed/_functional_collectives.py#L965

fegin reviewed (commented):
Thanks for fixing the issue. Is the PR stack ready to verify? If so, I would like to apply it to CP to see if the compilation issue still exists. Thanks!

IvanKobzarev commented:
> Thanks for fixing the issue. Is the PR stack ready to verify? If so, I would like to apply it to CP to see if the compilation issue still exists. Thanks!

We are iterating review-change on this PR, with applied review comments the last version is https://github.com/pytorch/pytorch/pull/132858

I see the inductor errors on CI, let me debug them first.

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

bdhirsh commented on torch/_higher_order_ops/effects.py:
```diff
@@ -123,6 +124,11 @@ def get_effect_key(op, args, kwargs) -> Optional[_EffectType]:
 
+def new_token_tensor() -> torch.Tensor:
+    # Use dtype bool to not affect Inductor dtype promotions
+    return torch.tensor([], dtype=torch.bool)
```
hmm so if we leave the token's dtype as `float32`, inductor might try to do some type promotion and bitcast the token, causing it to end up in a runtime triton kernel?

I would have (naively) expected the token to be unused anywhere in the compute graph in inductor IR, and should (therefore) not show up in any kernels. Do you have a code snippet showing where leaving the token's type as `float32` breaks?

IvanKobzarev replied:
Inductor lowering process before every lowering will `transform_args()`  https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L241
It will pick the highest dtype of all arguments, as a result of transform_args of with_effects(token, op, args) it will also count token.dtype and call promote for all args if their dtype is lower, e.g. args contain int32 dtype.

As a result it will create PointwiseOperation()

The dtype is picked here : https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L182

Technically we can improve this function to not count empty tensors.

bdhirsh replied:
I'm ok with this as long as @eellison is. Although if we do then we should probably change all of the sites in this file that create tokens to use `torch.bool` too: https://github.com/pytorch/pytorch/blob/main/torch/_higher_order_ops/effects.py#L219

IvanKobzarev commented:
> Thanks for fixing the issue. Is the PR stack ready to verify? If so, I would like to apply it to CP to see if the compilation issue still exists. Thanks!

@fegin Hi, the PR is ready to verify that it solves the original issues :)

fegin commented:
Please remove the two side effect fx code in , [main/torch/distributed/_functional_collectives.py#L965](https://github.com/pytorch/pytorch/blob/main/torch/distributed/_functional_collectives.py?rgh-link-date=2024-08-06T16%3A45%3A22Z#L965)

yifuwang commented on test/distributed/test_c10d_functional_native.py:
```diff
@@ -565,14 +565,14 @@ def func(args: List[torch.Tensor]) -> torch.Tensor:
-            .check("torch.ops._c10d_functional.wait_tensor.default(buf6")
+            .check("torch.ops._c10d_functional.wait_tensor.default(buf15")
+            .check("torch.ops._c10d_functional.wait_tensor.default(buf16")
+            .check("torch.ops._c10d_functional.wait_tensor.default(buf15")
+            .check("torch.ops._c10d_functional.wait_tensor.default(buf16")
```
This doesn't look right. Wait on buf0 and buf1 should take place.

IvanKobzarev replied:
There are some additional manipulations with buffers as a result of additional getitem() ops, as with_effects() returns tuple.
But to my understanding it should not incur additional copies.
Will be good of course to get rid of those additional assignments.

yifuwang replied:
Oh I wasn't worried about the aesthetics. I was saying that the test doesn't verify buf0 and buf1 are waited, which are required to ensure correctness.

yifuwang commented on test/distributed/test_c10d_functional_native.py:
```diff
@@ -565,14 +565,14 @@ def func(args: List[torch.Tensor]) -> torch.Tensor:
             # Expect no in-place with graph input (buf5, buf6 are clones)
```
Please update the buffer names in the comment.

yifuwang commented on test/distributed/test_c10d_functional_native.py:
```diff
@@ -833,16 +833,16 @@ def func(arg: torch.Tensor) -> torch.Tensor:
-            .check("torch.ops._c10d_functional.wait_tensor.default(buf0")
+            .check("torch.ops._c10d_functional.broadcast_.default(buf2")
+            .check("torch.ops._c10d_functional.wait_tensor.default(buf2")
             # Expect no in-place with graph input (buf5 is a clone)
```
It was already stale before your chagne, but can you update the buffer name in the comment?

yifuwang commented on test/distributed/_tensor/test_dtensor_compile.py:
```diff
@@ -557,11 +557,14 @@ def fn(x):
-    sin_1 = torch.ops.aten.sin.default(sin);  sin = None
-    return (sin_1, primals_1, wait_tensor)""",
+def forward(self, primals_1, primals_2):
+    with_effects = torch.ops.higher_order.with_effects(primals_1, torch.ops._c10d_functional.wait_tensor.default, primals_2);  primals_1 = None
+    getitem = with_effects[0]
+    getitem_1 = with_effects[1];  with_effects = None
```
Noob q: why does this take and return two arguments? Which one is the wait target/result? What is the other one?

IvanKobzarev replied:
The main idea of additional tokens is to wrap effectful ops into with_effects and add additional first argument that consumes "token" which is originally added to the graph inputs; with_effects will return (new_token, result), so with_effects[0] here is the next token, with_effects[1] is actual result:
```
def fn(x):
    return wait_tensor(x)
```
```
forward(token, x)
    new_token, y = with_effects(token, wait_tensor, x)
return new_token, y
```

yifuwang replied:
Thanks for the explanation! Just to verify my understanding, each graph will have at most one effect token as argument, which "chains" all the effectful ops?

bdhirsh replied:
that's right (in theory, you could have multiple effect token arguments to the graph, so you can create separate "dependency chains", if you decide that having on giant dependency chain for all effectful ops is too restrictive. Multiple tokens is not yet implemented though)

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

yifuwang commented on torch/_inductor/fx_passes/micro_pipeline_tp.py:
```diff
@@ -546,16 +662,21 @@ def fuse_all_gather_matmul(all_gather: _AllGatherMatch) -> None:
 
     # Find consumer matmuls
-    matmuls = _find_consumer_matmuls(ag_res_node)
+    if with_effects:
+        with_effects_output_token, with_effects_result = with_effects_users(
+            with_effects
+        )
+        if ag_res_node == with_effects:
+            ag_res_node = with_effects_result
```
This logic should be in `_find_consumer_matmuls`. We want to isolate `all_gather` pattern matching and the operator fusion, so that future changes to the all-gather graph patterns can be isolated.

yifuwang commented on torch/_inductor/fx_passes/micro_pipeline_tp.py:
```diff
@@ -575,6 +696,13 @@ def fuse_all_gather_matmul(all_gather: _AllGatherMatch) -> None:
             graph, matmuls, shard_node, gather_dim, group_name
         )
+
+        if with_effects:
```
ditto

yifuwang commented on torch/_inductor/fx_passes/micro_pipeline_tp.py:
```diff
@@ -590,7 +718,13 @@ def fuse_all_gather_matmul(all_gather: _AllGatherMatch) -> None:
             matmul.erase()
-        all_gather.replace_with(new_ag_node)
+
+        if with_effects:
```
This logic should be in all_gather.replace_with

yifuwang commented on torch/_inductor/fx_passes/micro_pipeline_tp.py:
```diff
@@ -523,19 +639,19 @@ def fuse_all_gather_matmul(all_gather: _AllGatherMatch) -> None:
     )
 
-    shard_node, ag_node, ag_res_node, gather_dim, group_name = (
+    shard_node, ag_node, ag_res_node, gather_dim, group_name, with_effects = (
```
In this function, we don't care about the nodes between the `ag_node` (node whose target is all-gather) and the `ag_res_node`. Let's encapsulate the `with_effects` handling in `_find_consumer_matmuls()` and `_AllGatherMatch.replace_with()`..

yifuwang commented:
Thanks for adapting the all-gather/reduce-scatter pattern matchers for the change! The general approach looks great. A few comments to make the pattern handling better encapsulated so that future changes like this are easier.

fegin commented on torch/_inductor/fx_passes/ddp_fusion.py:
```diff
@@ -95,16 +103,13 @@ def get_comm_block(comm_node: fx.Node) -> Optional[CommBlock]:
-        len(comm_node.users) == 1
-        and first_user.target == torch.ops._c10d_functional.wait_tensor.default
-    ):
+    if len(comm_node.users) == 1 and is_wait_tensor(first_user):
```
Thanks for the update. But we also need to update the call_function below that replace multiple wait_nodes with just one wait_node. This test is green before this PR: `pytest test/distributed/_composable/test_replicate_with_compiler.py  --verbosity=1 -k test_bucketing_concat_op`.

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented on torch/_inductor/lowering.py:
```diff
@@ -6245,6 +6245,10 @@ def _sink_tokens(tokens):
 def with_effects(token, op, *args, **kwargs):
+    if op == torch.ops._c10d_functional.wait_tensor.default:
+        assert len(args) == 1
+        return token, _wait_tensor(args[0])
```
Sorry, why do need to do this ?

eellison commented on torch/_inductor/fx_passes/ddp_fusion.py:
```diff
@@ -80,6 +84,17 @@ class CommBlock:
+
+def is_wait_tensor(node: torch.fx.Node) -> bool:
+    return (
+        node.op == "call_function"
+        and node.target == torch.ops._c10d_functional.wait_tensor.default
```
nit: you can remove `node.op == "call_function"` and just test `node.target == torch.ops._c10d_functional.wait_tensor.default`

eellison reviewed (commented):
Seems reasonable but a couple questions.

> Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

Where is this happening ? If it's inside `register_lowering(torch.ops.higher_order.with_effects)` I think we should just pass in `type_promotion_kind=None`

IvanKobzarev commented:
> Seems reasonable but a couple questions.
> 
> > Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32
> 
> Where is this happening ? If it's inside `register_lowering(torch.ops.higher_order.with_effects)` I think we should just pass in `type_promotion_kind=None`

The dtype promotion happens in inductor/lowering.py transform_args in before specific lowering handling

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L287

eellison commented:
That only occurs if `type_promotion_kind` is not None. which lowering is type promoting ?

IvanKobzarev commented:
> Why do we need this?

Added comment and created issue:
```
    # with_effects wrap result of the op in tuple.
    # As a result ir produces redundant no-op buffer assignments.
    # They are not affecting performance, but confusing for the tests,
    # that verify, that wait_tensor was called on exactly the result buffer of functional collectives.
    # E.g.:
    # buf14 = torch.ops._c10d_functional.wait_tensor.default(buf13)
    # del buf12
    # del buf13
    # buf15 = buf14
    # return buf15
    #
    # TODO(ivankobzarev): After optimization of those redundant assignments this shortcircuit can be remvoved
    # https://github.com/pytorch/pytorch/issues/136012
```

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev commented:
> That only occurs if `type_promotion_kind` is not None. which lowering is type promoting ?

Thanks.
If to set 
```
@register_lowering(torch.ops.higher_order.with_effects, type_promotion_kind=None)
```
We do not need to set tokens dtype=bool, will create a separate PR for this

IvanKobzarev committed:
Update on "Make c10d_functional.wait_tensor effectful ORDERED op"


Making wait_tensor effectful op.

- Removing previous solution to mark having side_effect for fx, that prevented from DCE with corresponding tests.

Inductor promotes arguments dtypes to the highest dtype, as a result additional token tensor argument wtih float32 dtype incurred dtype promotions for lower types, e.g. int32

The solution for that - to use the lowest dtype for tokens - torch.bool.
Alternative will be to exclude from dtype calculations zero-size tensors.

Making wait_tensor effectful adds more manipulations with buffers, as we wrapping into with_effects operation that returns tuple. As a result tests that were checking buffer numbers have changes on those buffers,

Adding short circuit for inductor lowering for with_effects(wait_tensor) to wait_tensor lowering, avoids most of those buffer manipulations.


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kwen2501 commented:
cc @yf225

yf225 commented on test/fx/test_dce_pass.py:
```diff
@@ -233,63 +232,3 @@ def forward(self, a: torch.Tensor) -> torch.Tensor:
-                future = torch.ops._c10d_functional.all_reduce.default(e, "sum", "0")
-                synced_e = torch.ops._c10d_functional.wait_tensor.default(
-                    future
-                )  # synced_e is not used
```
Curious why do we need to remove these two tests in this file?

The motivation for this test case is that even if rank 0 doesn't use the `synced_e` (all-reduce output), rank 1 might use their `synced_e`, and hence we still cannot DCE the all_reduce and wait_tensor in rank 0

IvanKobzarev replied:
This was testing fx tracing preservation in graph, relying on 
```
torch/distributed/_functional_collectives.py
-    # mark these ops has side effect so that they won't be removed by DCE
-    torch.fx.node.has_side_effect(torch.ops._c10d_functional.wait_tensor.default)
-    torch.fx.node.has_side_effect(torch.ops._c10d_functional.wait_tensor)
```
making wait_tensor effectful op will keep it from DCE for dynamo tracing only.

Potentially if fx case is needed - we can keep it.

yifuwang commented on torch/_inductor/fx_passes/micro_pipeline_tp.py:
```diff
@@ -69,8 +70,21 @@ class _AllGatherMatch:
+    with_effects: Optional[torch.fx.Node] = None
 
     def replace_with(self, new_node: torch.fx.Node) -> None:
+        if self.with_effects:
+            graph = new_node.graph
+            with graph.inserting_before(new_node):
+                make_token = graph.call_function(torch.ops.prims._make_token.default)
+            with_effects_out_token, with_effects_result = get_with_effects_users(
+                self.with_effects
+            )
+            with_effects_out_token.replace_all_uses_with(make_token)
+            graph.erase_node(with_effects_out_token)
+            with_effects_result.replace_all_uses_with(new_node)
+            graph.erase_node(with_effects_result)
+
```
Wow this looks painful :)

yifuwang commented:
It seems that after this PR, `a.replace_all_uses_with(b)` becomes complicated  when `a` is an effectful node. The PR modified a few instances of `.replace_all_uses_with()` and they all seem slightly different. As someone who would write this, I'm not confident that I wouldnt't mess up. Questions:

- The `.replace_all_uses_with` instances you modified checks whether the replacement target is an effectful node. Have we considered the possibility that passes that are not effectful-aware can also perform such replacements?
- Whenever someone mistakenly replaces an effectful op with `.replace_all_uses_with`, will there be an explicit error? Will `graph.lint()` catch the issue? If so, can we have some test on it?
- Can we abstract this logic into some util? In fact, since it seems that effectful token is going to be a first-class concept in fx, can we subsume the logic into `.replace_all_uses_with`?

<img width="1806" alt="image" src="https://github.com/user-attachments/assets/aae17632-9a4d-4953-8731-f317ae3cff2c">

<img width="1658" alt="image" src="https://github.com/user-attachments/assets/0aa3c4d8-f7e4-4490-9fa5-e08522b1b27e">

<img width="1633" alt="image" src="https://github.com/user-attachments/assets/2def1b0e-8c5a-4d95-9243-82a10cadb5ef">
